### ISSUE TITLE
Speed up

### ISSUE THREAD

**Issue by jonthegeek** [macmillancontentscience/wordpiece#27]
This package is slow when wordpiece-tokenizing more than a handful of examples. It's never been our rate-limiting step, so we didn't really notice. 

I tried memoising to improve, but it only sped things up for repeated full phrases (and actually made things slower when I memoised the single-word step, which I would have thought would have helped the most).

The best indicator of what is likely speed-uppable is that the memory allocation is ~5GB for tokenizing 120k of phrases. That feels like something blowing up in a loop. When I memoise the .wp_tokenize_word piece the allocation goes down quite a bit, which supports the idea that it's happening there.

A couple things to try:

- Sometimes pre-allocating vectors can help. We're largely doing that, so... this probably isn't it.
- Emil pointed out "stringi::stri_sub() is a vectorized version of substr() ". I haven't conceptualized how we could vectorize the while, but maybe that could work?

My best guess right now is to maybe try building all the possible substrings and checking for those, all at once, and then using that to construct the result. It doesn't seems like it would DEFINITELY be faster but it's the kind of thing R might work with better. But I tried a version of that and it was slower so... probably not. It still had a `while (start <= end)` loop though so it could maybe be improved. Notably I was able to eliminate the inner while, and I had to unlist at one point, so that indicates I was basically un-vectorizing.

I could see `purrr::accumulate` possibly being helpful here but so far... I don't see how.

So basically... do a better job than I did at learning to use the profiler to find slow points, I think!

Oh, this is the dataset Emil was using, which has about 1k unique labels. It really isn't that big but it's slow enough to notice, so we should fix that:

```r
library(modeldata)
data(tate_text)
library(wordpiece)
to_tokenize <- as.character(unique(tate_text$medium))

testing <- wordpiece_tokenize(to_tokenize)
```
